Mathlib/RingTheory/MvPolynomial/Symmetric.lean

```diff
@@ -309,4 +309,244 @@ theorem psum_isSymmetric (n : ℕ) : IsSymmetric (psum σ R n) := rename_psum _
 
 end PowerSum
 
+section Exponent
+
+open Classical
+open Composition
+open Finset
+
+variable (σ R) [CommSemiring R] [CommSemiring S] [Fintype σ] [Fintype τ]
+
+@[ext]
+structure Exponent (n : ℕ) :=
+  (exp : σ → ℕ)
+  (sum_eq : ∑ i, exp i = n)
+
+lemma Exponent.single_le_sum {n : ℕ} (s : Exponent σ n) (i : σ) : s.exp i ≤ n := by
+  have h : ∑ i, s.exp i = n := s.sum_eq
+  simp_rw [←h]
+  exact Finset.single_le_sum (fun i _ => Nat.zero_le (s.exp i)) (mem_univ i)
+  done
+
+lemma Exponent.zero_of_sum_eq_zero (s : Exponent σ 0) : ∀ i, s.exp i = 0 := by
+  intro i
+  have := Finset.single_le_sum (fun i _ => Nat.zero_le (s.exp i)) (mem_univ i)
+  rw [s.sum_eq] at this
+  exact Nat.eq_zero_of_le_zero this
+  done
+
+def Exponent.zero : Exponent σ 0 := ⟨fun _ => 0, by norm_num⟩
+
+def Exponent.one (i : σ) : Exponent σ 1 := ⟨fun j => if i = j then 1 else 0, by norm_num⟩
+
+instance exponentFinite (n : ℕ) : Finite (Exponent σ n) := by
+  let f : Exponent σ n → (σ → Fin (n+1)) := fun s => fun i => ⟨s.exp i, Nat.lt_succ_of_le (Exponent.single_le_sum σ s i)⟩
+  have hf : Function.Injective f := by
+    intros s t h
+    rw [Exponent.ext_iff]
+    ext i
+    have := congr_arg (fun s => s i) h
+    simp only [Function.Injective.eq_iff] at this
+    exact Fin.mk_eq_mk.mp this
+  exact Finite.of_injective f hf
+  done
+
+instance exponentFintype (n : ℕ) : Fintype (Exponent σ n) := Fintype.ofFinite (Exponent σ n)
+
+def exponentLiftEquiv (e : σ ≃ τ) : Exponent σ n ≃ Exponent τ n := {
```

Comment:
A few comments, just about the style
- just two spaces of indentation will suffice
- I think (untested) you can write these as `toFun s := ` etc.
- you can use dot notation, e.g. replace `Equiv.sum_comp e` by `e.sum_comp`.